llvm/lib/Transforms/FuzzIntrospector/FuzzIntrospector.cpp

```diff
@@ -1132,3 +1184,234 @@ static RegisterStandardPasses
+}
+
+// Calculate complexities reachable from each reachable unique BBs
+// TODO: for now, this function just accounts for complexity of callee functions from 
+// the reachable BBs. We can do better by carefully calculating complexity of the reachable
+// regions of the code i.e. intra-procedural complexity by counting the number of unique 
+// edges and nodes for each reachable set.
+std::pair<size_t, size_t> FuzzIntrospector::findComplexities(SmallPtrSet<BasicBlock *, 32> trueReachable, SmallPtrSet<BasicBlock *, 32> falseReachable, std::map<BasicBlock *, size_t> BBComplexityMap) {
+  size_t trueComp = 0, falseComp = 0;
+
+  // iterate and skip those reachable by false side
+  for (auto BB: trueReachable) {
+    if (falseReachable.find(BB) != falseReachable.end()) {
+      continue;
+    }
+
+    if (BBComplexityMap.find(BB) == BBComplexityMap.end()) {
+      BBComplexityMap[BB] = calculateBBComplexity(BB);
+    }
+    trueComp += BBComplexityMap[BB];
+  }
+
+  // iterate and skip those reachable by true side
+  for (auto BB: falseReachable) {
+    if (trueReachable.find(BB) != trueReachable.end()) {
+      continue;
+    }
+
+    if (BBComplexityMap.find(BB) == BBComplexityMap.end()) {
+      BBComplexityMap[BB] = calculateBBComplexity(BB);
+    }
+    falseComp += BBComplexityMap[BB];
+  }
+
+  return make_pair(trueComp, falseComp);
+}
+
+std::pair<std::string, std::string> FuzzIntrospector::getInsnDebugInfo(Instruction *I) {
+  std::string ret_string = "", ret_line = "";
+
+  DILocation *Loc = I->getDebugLoc();
+  if (Loc != NULL)
+  {
+    ret_line = std::to_string(Loc->getLine());
+    ret_string = Loc->getFilename().str() +
+              +":" + ret_line + "," + std::to_string(Loc->getColumn());
+  }
+  else {
+    logPrintf(L1, "No debug info!!\n");
+  }
+
+  return make_pair(ret_string, ret_line);
+}
+
+std::pair<std::string, std::string> FuzzIntrospector::getBBDebugInfo(BasicBlock *BB, DILocation *prevLoc) {
+  std::pair<std::string, std::string> result = make_pair("", "");
+
+  BasicBlock *currBB = BB;
+  BranchInst *currBI;
+  Instruction *currTI, *currI;
+  DILocation *currLoc;
+
+  do {
+    currTI = currBB->getTerminator();
+    currI = currBB->getFirstNonPHIOrDbgOrLifetime(true);
+    if (currI == nullptr)
+      break;
+    currLoc = currI->getDebugLoc();
+    currBI = dyn_cast<BranchInst>(currTI);
+    if (currBI && !currBI->isConditional()){
+      currBB = currBI->getSuccessor(0);
+    }
+    else
+      break;
+  } while (currLoc == prevLoc);
+
+  if (currI)
+    result = getInsnDebugInfo(currI);
+  // for (auto &I: *BB) {
+  //   result = getInsnDebugInfo(&I);
+  //   if (result.first.length() > 0) {
+  //     break;
+  //   }
+  // }
+  return result;
+}
+
+// void FuzzIntrospector::writeOutMap(std::map<std::string, std::vector<std::pair<std::string, size_t>>> outMap, std::string fileName) {
+// void FuzzIntrospector::writeOutMap(std::map<std::string, BranchSidesComplexity> outMap, std::string fileName) {
```

Comment:
Do we want to keep these two lines? I'm okay with leaving comments like this in it helps your dev process, but if they are not needed for future work I suggest removing

Reply:
removed.